### PR TITLE
add polling to claims status

### PR DESCRIPTION
**Pull request by rianfowler**
[Ticket](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/12278). Will put appeals status polling in another PR.

### CODE REVIEW

File: src/applications/claims-status/actions/index.jsx
Comment (rianfowler) [inline src/applications/claims-status/actions/index.jsx]: Noticed a lot of our polling implementations do timeout-first polling. Set up a poll first approach here.

File: src/applications/claims-status/actions/index.jsx
Comment (rianfowler) [inline src/applications/claims-status/actions/index.jsx]: This obviates the need to `mockFetch` which is tedious and pointless when we have a fetch helper that's already unit tested.

File: src/applications/claims-status/tests/actions.unit.spec.js
Comment (rianfowler) [inline src/applications/claims-status/tests/actions.unit.spec.js]: Unit testing recursive polling is gross.

**Review by cehsu — APPROVED** [review]
LGTM, pending a response to the question regarding what to dispatch if anything on success.

File: src/applications/claims-status/actions/index.jsx
Comment (rianfowler) [reply src/applications/claims-status/actions/index.jsx]: should we be using the get from the platform helpers?

File: src/applications/claims-status/actions/index.jsx
Comment (rianfowler) [reply src/applications/claims-status/actions/index.jsx]: I will add this to the clean up ticket I am going to make to remove the old claims status. We need some more information on what's returned on a failure and how it should be displayed. 

This approach just wraps the existing [behavior](https://github.com/department-of-veterans-affairs/vets-website/blob/5db1378e4b27c0b9a7ed95250aab8d50dd630fab/src/applications/claims-status/actions/index.jsx#L157) with polling.

File: src/applications/claims-status/actions/index.jsx
Comment (cehsu) [reply src/applications/claims-status/actions/index.jsx]: Good idea.

File: src/applications/claims-status/actions/index.jsx
Comment (rianfowler) [reply src/applications/claims-status/actions/index.jsx]: Not really sure what the best composition is. The polling logic is just `setTimeout` with a recursive call but I was thinking it would be good to have the forking behavior grouped i.e. 
- does it poll again?
- does it succeed?
- does it fail? 

I could pass another predicate / callback pair in like `shouldPoll` and `onPoll` so it's more consistent. Then I wouldn't need to pass in the polling interval. 

screen shot 2018-08-17 at 12 11 23

File: src/applications/claims-status/actions/index.jsx
Comment (rianfowler) [reply src/applications/claims-status/actions/index.jsx]: Yes 
https://app.zenhub.com/workspace/o/department-of-veterans-affairs/vets.gov-team/issues/12469

File: src/applications/claims-status/actions/index.jsx
Comment (rianfowler) [reply src/applications/claims-status/actions/index.jsx]: Okay I see what you mean- I'll refactor this a bit.

Comment (rianfowler) [reply src/applications/claims-status/actions/index.jsx]: Are you talking about drying the `get`? 

```
shouldFail: response => _.get(response, 'meta.syncStatus') === 'FAILED',
shouldSucceed: response => _.get(response, 'meta.syncStatus') === 'SUCCESS'
```

Comment (cehsu) [reply src/applications/claims-status/actions/index.jsx]: Precisely.

Comment (rianfowler) [reply src/applications/claims-status/actions/index.jsx]: Refactored that get to use the platform and added a helper for that.

I tried a couple of different ways to reorganize this but decided to leave it mostly the same. Couple differences from the burials polling example:
- It takes a timeout first approach which means the initial request doesn't fire until the first polling timeout 
- It's not using the `apiRequest` helper which has our error reporting baked in and is not promisified

File: src/applications/claims-status/actions/index.jsx
Comment (cehsu) [inline src/applications/claims-status/actions/index.jsx]: Will we also be making use of the `show` (`/v0/evss_claims_async/:id`) endpoint mentioned in the linked issue in another PR?

File: src/applications/claims-status/actions/index.jsx
Comment (cehsu) [inline src/applications/claims-status/actions/index.jsx]: Do we anticipate using this in a setting where we would want to be able to configure the pollingInterval, rather than accessing it as a variable in the outer scope?

**Review by cehsu — APPROVED** [review]
LGTM. I had a few minor organizational questions for your consideration.

File: src/applications/claims-status/actions/index.jsx
Comment (cehsu) [inline src/applications/claims-status/actions/index.jsx]: Do we anticipate wanting to configure shouldFail and shouldSucceed? If not, could we pull this logic into a checkStatus helper to house these checks, as well as the call to pollClaimsStatus (Similar to the burial app)?

**Review by cehsu — APPROVED** [review]
LGTM. Meant to include this question in the previous review.

File: src/applications/claims-status/actions/index.jsx
Comment (rianfowler) [reply src/applications/claims-status/actions/index.jsx]: I plan on reusing this for the appeals status. Could have gone with a switch statement- this is mostly a style decision on my part to invert the control so it's easier to tell what's going on without having to read the polling function. I figure a maintainer will get to this from the action name or the action creator call, and they can tell what this is doing just by reading the arguments. 

screen shot 2018-08-20 at 13 53 04

File: src/applications/claims-status/actions/index.jsx
Comment (rianfowler) [reply src/applications/claims-status/actions/index.jsx]: Nope- I would rather use the hard coded value since it's not reused. I'll update this.

File: src/applications/claims-status/actions/index.jsx
Comment (cehsu) [reply src/applications/claims-status/actions/index.jsx]: Thanks for the context, sounds good.

Comment (cehsu) [reply src/applications/claims-status/actions/index.jsx]: Since we are planning to reuse this for appeals, will it have the same shouldFail, and shouldSucceed logic? If so, could that simply be specified, rather than require configuration?

Comment (rianfowler) [reply src/applications/claims-status/actions/index.jsx]: I think the extra boilerplate is fine- I can read the `poll` function's arguments and tell what its doing without reading its implementation. If I push the predicates into the polling function then I know it's "polling" but I don't know when it's going to call success or failure without looking at its implementation. 

Dunno if this is a legitimate design approach, though- it's inversion of control to me.

**Pull request merged**
